crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1620,37 +1626,49 @@ impl<'db> KnownClass {
 
 /// Enumeration of specific runtime that are special enough to be considered their own type.
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
-pub enum KnownInstanceType {
+pub enum KnownInstanceType<'db> {
     /// The symbol `typing.Literal` (which can also be found as `typing_extensions.Literal`)
     Literal,
+    /// A single instance of `typing.TypeVar`
+    TypeVar(TypeVarInstance<'db>),
```

Comment:
> I think as we build out support for type variables across our data model the differences to the other variants of this enum will only continue to grow.

I think this is the point that I don't believe is true, which gives me a different perspective on this tradeoff.

When we support typevars as annotations, we will need to add a representation for "something typed by a type variable", e.g. `x: T`. This is a very special type, and we will definitely need a dedicated `Type` variant for it. It certainly will not reuse the type of the type variable object itself (though it will likely internally store a reference to the same `TypeVarInstance`.) The type of the type variable itself is a singleton type representing exactly one instance of `typing.TypeVar`, whereas the type of `x` in `x: T` is the type "some unknown set of objects, constrained by the upper bound or constraints on the type var." Reusing the same type variant for these would be similar to trying to share the same `Type` variant for `Type::ClassLiteral` and `Type::Instance` -- they are too semantically different for that to make sense.

The type of the type variable itself, as a runtime object, which is what we have in this PR, is (I believe) feature-complete as of this PR, and shouldn't need to change or expand (though of course we will need to add paramspecs and typevartuples). It is nothing more than a known instance with some properties whose type we can infer more precisely than typeshed. Typevar objects don't have a lot of interesting behavior in themselves, so I don't see what else we would need to add.

I think it's quite a significant benefit to use `KnownInstance` for this type to get the fallback behavior for free, and avoid the need to add another variant, with very similar behavior to `KnownInstance`, to every Type match.

> But they differ in that there is only one `typing.Literal` symbol and only one `typing.NoDefault` symbol, whereas there are many `TypeVar` instances.

There is exactly one instance of each `TypeVar`, though. Which is why the `TypeVar` variant holds data to describe _which_ typevar it represents, but the `typing.NoDefault` and `typing.Literal` variants don't. It is still the case that every `Type::KnownInstance` represents exactly one singleton object.

Reply:
Ahhh, I see. Okay, this makes a lot more sense. Please add a doc-comment to the `TypeVarInstance` struct clarifying this, though! I did indeed misunderstand the purpose of this variant and that struct when reading through the PR.